diff -ruN a/xl2tpd.c b/xl2tpd.c
--- a/xl2tpd.c	2018-01-02 15:45:27.000000000 +0800
+++ b/xl2tpd.c	2018-02-26 16:26:54.092012818 +0800
@@ -467,11 +467,13 @@
 
 #ifdef DEBUG_PPPD
     l2tp_log (LOG_DEBUG, "%s: I'm running: \n", __FUNCTION__);
+#if 0
     for (x = 0; stropt[x]; x++)
     {
         l2tp_log (LOG_DEBUG, "\"%s\" \n", stropt[x]);
     };
 #endif
+#endif
 #ifdef __uClinux__
     c->pppd = vfork ();
 #else 
@@ -1523,13 +1525,15 @@
 
     open_controlfd();
 
-    l2tp_log (LOG_INFO, "xl2tpd version " SERVER_VERSION " started on %s PID:%d\n",
+    l2tp_log (LOG_INFO, "xl2tpd started on %s PID:%d\n",
          hostname, getpid ());
+#if 0
     l2tp_log (LOG_INFO,
          "Written by Mark Spencer, Copyright (C) 1998, Adtran, Inc.\n");
     l2tp_log (LOG_INFO, "Forked by Scott Balmos and David Stipp, (C) 2001\n");
     l2tp_log (LOG_INFO, "Inherited by Jeff McAdams, (C) 2002\n");
     l2tp_log (LOG_INFO, "Forked again by Xelerance (www.xelerance.com) (C) 2006\n");
+#endif
     listenaddr.s_addr = gconfig.listenaddr;
     l2tp_log (LOG_INFO, "Listening on IP address %s, port %d\n",
               inet_ntoa(listenaddr), gconfig.port);
